File: packages/typescript/src/pin/render.ts
/**
 * Turn an {@link UpgradeReport} into the bytes to write.
 *
 * Pure: no console I/O, no CLI types, no exceptions on well-formed input. The
 * CLI keeps sole ownership of *which stream* each piece of output goes to and
 * of exit codes (ADR-0007); this module owns *what the bytes are*.
 *
 * Everything a caller used to have to know about producing `deps upgrade`
 * output lives here now: the four formats and their exact bytes, the
 * trailing-newline discipline, the empty-report shape per format, and the
 * mapping from "what the engine looked for" to "which JSON keys appear".
 */

import type { LockfileStaleEntry, UpgradeReport, VersionBump } from "./engine.js";

/** The output shapes {@link renderUpgradeReport} can produce. */
export type UpgradeFormat = "text" | "json" | "pr-body" | "issue-body";

/**
 * Render an upgrade report, returning the exact bytes to write.
 *
 * Pure: no console I/O, no CLI types. The returned string is already
 * terminated as it should be written — the caller writes it verbatim and
 * neither adds nor suppresses a trailing newline.
 *
 * Which JSON keys appear is decided by `report.checkedVersions` /
 * `report.checkedLockfile`, never re-derived from the CLI's `--mode`. That
 * holds for empty reports too: a run asked only for version tags emits only
 * `version_bumps`, whether or not it found any.
 *
 * @param report - The typed outcome of an {@link upgrade} run.
 * @param outputFormat - One of {@link UpgradeFormat}. An unrecognised value is
 *   a programmer error — the CLI validates `--format` before calling.
 * @throws Error if `outputFormat` is not a known format.
 */
export function renderUpgradeReport(
  report: UpgradeReport,
  outputFormat: UpgradeFormat = "text",
): string {
  switch (outputFormat) {
    case "json":
      return renderJson(report);
    case "pr-body":
      return renderPrBody(report.versionBumps, report.lockfileStale);
    case "issue-body":
      return renderIssueBody(report.versionBumps, report.lockfileStale);
    case "text":
      return renderText(report.versionBumps, report.lockfileStale);
    default: {
      const unknown: never = outputFormat;
      throw new Error(`unknown output format ${JSON.stringify(unknown)}`);
    }
  }
}

/**
 * Render `--format json`, keyed by what the run was asked to check.
 *
 * Key presence follows `checkedVersions` / `checkedLockfile` in *all* cases,
 * empty included. This resolves the open choice recorded in
 * `docs/specs/0005-typed-engine-report-seam.md` §2.2 the other way: the key set
 * now depends only on `--mode`, which the caller chose, and not on whether the
 * run happened to find anything.
 */
function renderJson(report: UpgradeReport): string {
  const result: {
    version_bumps?: Array<Record<string, unknown>>;
    lockfile_stale?: Array<Record<string, unknown>>;
  } = {};
  if (report.checkedVersions) {
    result.version_bumps = report.versionBumps.map(bumpToJson);
  }
  if (report.checkedLockfile) {
    result.lockfile_stale = report.lockfileStale.map(staleToJson);
  }
  return JSON.stringify(result, null, 2) + "\n";
}

/** Serialize a version bump for `--format json` (omitting empty `source_files`). */
function bumpToJson(bump: VersionBump): Record<string, unknown> {
  const entry: Record<string, unknown> = {
    uses: bump.uses,
    current: bump.current,
    latest: bump.latest,
    severity: bump.severity,
  };
  if (bump.source_files.length > 0) {
    entry.source_files = [...bump.source_files];
  }
  return entry;
}

/** Serialize a stale entry for `--format json` (omitting empty `source_files`). */
function staleToJson(stale: LockfileStaleEntry): Record<string, unknown> {
  const entry: Record<string, unknown> = {
    uses: stale.uses,
    current_sha: stale.current_sha,
    latest_sha: stale.latest_sha,
  };
  if (stale.source_files.length > 0) {
    entry.source_files = [...stale.source_files];
  }
  return entry;
}

/**
 * Render the pull-request body markdown for an upgrade report.
 *
 * Golden-file tested against `fixtures/expected/upgrade_pr_body.md` and kept
 * byte-identical with the Python port's `_render_pr_body`. An empty report
 * yields the bare header.
 */
function renderPrBody(versionBumps: VersionBump[], lockfileStale: LockfileStaleEntry[]): string {
  const lines: string[] = ["## ghagen dependency update", ""];

  if (versionBumps.length > 0) {
    lines.push("### Version bumps", "");
    for (const bump of versionBumps) {
      lines.push(`- \`${bump.uses}\` -> \`${bump.latest}\` [${bump.severity}]`);
    }
    lines.push("");
  }

  if (lockfileStale.length > 0) {
    lines.push("### Lockfile maintenance", "");
    for (const entry of lockfileStale) {
      lines.push(`- \`${entry.uses}\` SHA refreshed`);
    }
    lines.push("");
  }

  return lines.join("\n");
}

/**
 * Render the issue body markdown for an upgrade report.
 *
 * Golden-file tested against `fixtures/expected/upgrade_issue_body.md` and kept
 * byte-identical with the Python port's `_render_issue_body`. An empty report
 * yields the empty string.
 */
function renderIssueBody(versionBumps: VersionBump[], lockfileStale: LockfileStaleEntry[]): string {
  const lines: string[] = [];

  if (versionBumps.length > 0) {
    lines.push("## Version updates available", "");
    for (const bump of versionBumps) {
      let line = `- [ ] \`${bump.uses}\` -> \`${bump.latest}\` [${bump.severity}]`;
      if (bump.source_files.length > 0) {
        const files = bump.source_files.map((f) => `\`${f}\``).join(", ");
        line += `  in ${files}`;
      }
      lines.push(line);
    }
    lines.push("");
  }

  if (lockfileStale.length > 0) {
    lines.push("## Stale lockfile entries", "");
    lines.push("Run `ghagen deps pin --update` to refresh.", "");
    for (const entry of lockfileStale) {
      lines.push(`- [ ] \`${entry.uses}\` — SHA changed`);
    }
    lines.push("");
  }

  return lines.join("\n");
}

/**
 * Render the default human-readable report.
 *
 * Golden-file tested against `fixtures/expected/upgrade_text.txt` and kept
 * byte-identical with the Python port's `_render_text`. A report with nothing
 * to say yields the up-to-date line rather than the empty string — this is the
 * one format whose empty case has words in it.
 */
function renderText(versionBumps: VersionBump[], lockfileStale: LockfileStaleEntry[]): string {
  const lines: string[] = [];

  if (versionBumps.length > 0) {
    lines.push("Version updates available:", "");
    for (const bump of versionBumps) {
      lines.push(`  ${bump.uses}  →  ${bump.latest}  [${bump.severity}]`);
      for (const src of bump.source_files) {
        lines.push(`    in ${src}`);
      }
    }
    lines.push("");
  }

  if (lockfileStale.length > 0) {
    lines.push("Stale lockfile entries:", "");
    for (const entry of lockfileStale) {
      lines.push(`  ${entry.uses}`);
      lines.push(`    current SHA: ${entry.current_sha.slice(0, 12)}...`);
      lines.push(`    latest SHA:  ${entry.latest_sha.slice(0, 12)}...`);
    }
    lines.push("");
  }

  if (lines.length === 0) {
    return "Everything is up to date.\n";
  }

  return lines.map((line) => `${line}\n`).join("");
}

File: packages/python/src/ghagen/pin/render.py
"""Turn an :class:`~ghagen.pin.engine.UpgradeReport` into the bytes to write.

Pure: no console I/O, no CLI types, no exceptions on well-formed input.  The
CLI keeps sole ownership of *which stream* each piece of output goes to and of
exit codes (ADR-0007); this module owns *what the bytes are*.

Everything a caller used to have to know about producing ``deps upgrade``
output lives here now: the four formats and their exact bytes, the
trailing-newline discipline, the empty-report shape per format, and the
mapping from "what the engine looked for" to "which JSON keys appear".
"""

from __future__ import annotations

import json as json_mod
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from ghagen.pin.engine import LockfileStaleEntry, UpgradeReport, VersionBump

UpgradeFormat = Literal["text", "json", "pr-body", "issue-body"]
"""The output shapes :func:`render_upgrade_report` can produce."""


def render_upgrade_report(
    report: UpgradeReport, *, output_format: UpgradeFormat = "text"
) -> str:
    """Render an upgrade report, returning the exact bytes to write.

    Pure: no console I/O, no CLI types.  The returned string is already
    terminated as it should be written -- the caller writes it verbatim and
    neither adds nor suppresses a trailing newline.

    Which JSON keys appear is decided by ``report.checked_versions`` /
    ``report.checked_lockfile``, never re-derived from the CLI's ``--mode``.
    That holds for empty reports too: a run asked only for version tags emits
    only ``version_bumps``, whether or not it found any.

    Args:
        report: The typed outcome of an :func:`~ghagen.pin.engine.upgrade` run.
        output_format: One of :data:`UpgradeFormat`.  An unrecognised value is
            a programmer error -- the CLI validates ``--format`` before
            calling.

    Raises:
        ValueError: If *output_format* is not a known format.
    """
    if output_format == "json":
        return _render_json(report)
    if output_format == "pr-body":
        return _render_pr_body(report.version_bumps, report.lockfile_stale)
    if output_format == "issue-body":
        return _render_issue_body(report.version_bumps, report.lockfile_stale)
    if output_format == "text":
        return _render_text(report.version_bumps, report.lockfile_stale)
    raise ValueError(f"unknown output format {output_format!r}")


def _render_json(report: UpgradeReport) -> str:
    """Render ``--format json``, keyed by what the run was asked to check.

    Key presence follows ``checked_versions`` / ``checked_lockfile`` in *all*
    cases, empty included.  This resolves the open choice recorded in
    ``docs/specs/0005-typed-engine-report-seam.md`` Section 2.2 the other way:
    the key set now depends only on ``--mode``, which the caller chose, and not
    on whether the run happened to find anything.
    """
    result: dict = {}
    if report.checked_versions:
        result["version_bumps"] = [_bump_to_json(bump) for bump in report.version_bumps]
    if report.checked_lockfile:
        result["lockfile_stale"] = [
            _stale_to_json(entry) for entry in report.lockfile_stale
        ]
    return json_mod.dumps(result, indent=2) + "\n"


def _bump_to_json(bump: VersionBump) -> dict:
    """Serialize a version bump for ``--format json`` (omits empty ``source_files``)."""
    entry: dict = {
        "uses": bump.uses,
        "current": bump.current,
        "latest": bump.latest,
        "severity": bump.severity,
    }
    if bump.source_files:
        entry["source_files"] = list(bump.source_files)
    return entry


def _stale_to_json(stale: LockfileStaleEntry) -> dict:
    """Serialize a stale entry for ``--format json`` (omits empty ``source_files``)."""
    entry: dict = {
        "uses": stale.uses,
        "current_sha": stale.current_sha,
        "latest_sha": stale.latest_sha,
    }
    if stale.source_files:
        entry["source_files"] = list(stale.source_files)
    return entry


def _render_pr_body(
    version_bumps: list[VersionBump],
    lockfile_stale: list[LockfileStaleEntry],
) -> str:
    """Render the pull-request body markdown for an upgrade report.

    Golden-file tested against ``fixtures/expected/upgrade_pr_body.md`` and kept
    byte-identical with the TypeScript port's ``renderPrBody``.  An empty report
    yields the bare header.
    """
    lines = ["## ghagen dependency update", ""]

    if version_bumps:
        lines.append("### Version bumps")
        lines.append("")
        for bump in version_bumps:
            lines.append(f"- `{bump.uses}` -> `{bump.latest}` [{bump.severity}]")
        lines.append("")

    if lockfile_stale:
        lines.append("### Lockfile maintenance")
        lines.append("")
        for entry in lockfile_stale:
            lines.append(f"- `{entry.uses}` SHA refreshed")
        lines.append("")

    return "\n".join(lines)


def _render_issue_body(
    version_bumps: list[VersionBump],
    lockfile_stale: list[LockfileStaleEntry],
) -> str:
    """Render the issue body markdown for an upgrade report.

    Golden-file tested against ``fixtures/expected/upgrade_issue_body.md`` and
    kept byte-identical with the TypeScript port's ``renderIssueBody``.  An
    empty report yields the empty string.
    """
    lines: list[str] = []

    if version_bumps:
        lines.append("## Version updates available")
        lines.append("")
        for bump in version_bumps:
            line = f"- [ ] `{bump.uses}` -> `{bump.latest}` [{bump.severity}]"
            if bump.source_files:
                files = ", ".join(f"`{f}`" for f in bump.source_files)
                line += f"  in {files}"
            lines.append(line)
        lines.append("")

    if lockfile_stale:
        lines.append("## Stale lockfile entries")
        lines.append("")
        lines.append("Run `ghagen deps pin --update` to refresh.")
        lines.append("")
        for entry in lockfile_stale:
            lines.append(f"- [ ] `{entry.uses}` — SHA changed")
        lines.append("")

    return "\n".join(lines)


def _render_text(
    version_bumps: list[VersionBump],
    lockfile_stale: list[LockfileStaleEntry],
) -> str:
    """Render the default human-readable report.

    Golden-file tested against ``fixtures/expected/upgrade_text.txt`` and kept
    byte-identical with the TypeScript port's ``renderText``.  A report with
    nothing to say yields the up-to-date line rather than the empty string --
    this is the one format whose empty case has words in it.
    """
    lines: list[str] = []

    if version_bumps:
        lines.append("Version updates available:")
        lines.append("")
        for bump in version_bumps:
            lines.append(f"  {bump.uses}  →  {bump.latest}  [{bump.severity}]")
            for src in bump.source_files:
                lines.append(f"    in {src}")
        lines.append("")

    if lockfile_stale:
        lines.append("Stale lockfile entries:")
        lines.append("")
        for entry in lockfile_stale:
            lines.append(f"  {entry.uses}")
            lines.append(f"    current SHA: {entry.current_sha[:12]}...")
            lines.append(f"    latest SHA:  {entry.latest_sha[:12]}...")
        lines.append("")

    if not lines:
        return "Everything is up to date.\n"

    return "".join(f"{line}\n" for line in lines)
